src/lib/arch/aarch64/ExceptionHandler.cc

```diff
@@ -700,167 +189,74 @@ bool ExceptionHandler::init() {
                             ARM64_SYSREG_SVCR))});
     regValues.push_back(RegisterValue(newSVCR, 8));
-    instruction_.getArchitecture().setSVCRval(newSVCR);
+    instruction_->getArchitecture().setSVCRval(newSVCR);
 
-    ProcessStateChange stateChange = {ChangeType::REPLACEMENT, regs, regValues};
-    return concludeSyscall(stateChange);
+    simeng::OS::ProcessStateChange stateChange = {
+        simeng::OS::ChangeType::REPLACEMENT, regs, regValues};
+    processSyscallResult({false, 0, 0, stateChange});
+    return concludeSyscall();
   }
 
-  printException(instruction_);
+  printException();
   return fatal();
 }
 
-bool ExceptionHandler::readStringThen(char* buffer, uint64_t address,
-                                      int maxLength,
-                                      std::function<bool(size_t length)> then,
-                                      int offset) {
-  if (maxLength <= 0) {
-    return then(offset);
-  }
-
-  if (offset == -1) {
-    // First call; trigger read for address 0
-    memory_.requestRead({address + offset + 1, 1});
-    resumeHandling_ = [=]() {
-      return readStringThen(buffer, address, maxLength, then, offset + 1);
-    };
-    return false;
-  }
-
-  // Search completed memory requests for the needed data
-  bool found = false;
-  for (const auto& response : memory_.getCompletedReads()) {
-    if (response.target.address == address + offset) {
-      // TODO: Detect and handle any faults
-      assert(response.data && "Memory read failed");
-      buffer[offset] = response.data.get<char>();
-      found = true;
-      break;
-    }
-  }
-  memory_.clearCompletedReads();
-
-  if (!found) {
-    // Leave this handler in place to call again
-    return false;
-  }
-
-  if (buffer[offset] == '\0') {
-    // End of string; call onwards
-    return then(offset);
-  }
-
-  if (offset + 1 == maxLength) {
-    // Reached max length; call onwards
-    return then(maxLength);
-  }
-
-  // Queue up read for next character
-  memory_.requestRead({address + offset + 1, 1});
-  resumeHandling_ = [=]() {
-    return readStringThen(buffer, address, maxLength, then, offset + 1);
-  };
-  return false;
+void ExceptionHandler::processSyscallResult(
+    simeng::OS::SyscallResult syscallResult) {
+  syscallResult_ = syscallResult;
+  syscallReturned_ = true;
 }
 
-void ExceptionHandler::readLinkAt(span<char> path) {
-  if (path.size() == PATH_MAX_LEN) {
-    // TODO: Handle PATH_MAX_LEN case
-    std::cout << "\n[SimEng:ExceptionHandler] Path exceeds PATH_MAX_LEN"
-              << std::endl;
-    fatal();
-    return;
-  }
-
-  const auto& registerFileSet = core.getArchitecturalRegisterFileSet();
-  const auto dirfd = registerFileSet.get(R0).get<int64_t>();
-  const auto bufAddress = registerFileSet.get(R2).get<uint64_t>();
-  const auto bufSize = registerFileSet.get(R3).get<uint64_t>();
+bool ExceptionHandler::concludeSyscall() {
+  if (!syscallReturned_) return false;
```

Comment:
Is this check needed given there is a similar check in tick, and anywhere else concludeSyscall is called it is preceeded by processSyscallResult.

Reply:
Once again good spot, removing redundant check